tidy3d/components/simulation.py

```diff
@@ -2052,3 +2052,76 @@ def custom_datasets(self) -> List[Dataset]:
         """
         datasets = [src.field_dataset for src in self.sources if isinstance(src, CustomFieldSource)]
         return datasets
+
+    def volumetric_equivalent(self) -> Simulation:
+        """Generate new Simulation wherein any 2D materials are converted to 3D
+        volumetric equivalents."""
+
+        small_num = 1e-5
+
+        # get grid size at monolayer
+        def get_dl(sim, geom, axis):
+            dls = sim.discretize(geom).sizes.to_list[axis]
+            if len(dls) != 1:
+                log.error("2d materials do not support this nonuniform grid.")
+            return dls[0]
+
+        # snap monolayer to Yee grid
+        def snap_to_grid(sim, geom, axis):
+            new_center = list(geom.center)
+            new_center[axis] = sim.discretize(geom).boundaries.to_list[axis][0]
+            return geom.updated_copy(center=new_center)
+
+        # find neighboring material properties and grid sizes
+        def get_neighbors(sim, geom, axis, structures):
+            dl = get_dl(sim, geom, axis)
+
+            center1 = list(geom.center)
```

Comment:
feel like this stuff could be refactored a little for clarity and to remove some redundancy, eg

```py
center = list(geom.center)
neighboring_media = []
for up_down_index, dl_signed in enumerate((-dl, +dl)):
    center_shifted = list(geom.center)
    center_shifted[axis] += dl_signed * small_num
    geometry_shifted = geometry.updated_copy(center=center_shifted)
    media = sim.intersecting_media(geometry_shifted, structures)
    if len(media) > 1:
        raise SetupError("2D materials do not support multiple neighboring media on a side.")
    medium_edge = Medium() if len(media) == 0 else media[0]
    dl_edge = get_dl(sim, geometry_shifted, axis)
    neighboring_media.append((medium_edge, dl_edge))
return neighboring_media
```